File: nexus/ui/src/api/query/kinds.ts
import { fetchJson, readCsrfHeader } from "@nube/starter-client-ts";
import type { StarterClient } from "@nube/starter-client-ts";

import type {
  CreateQueryKindRequest,
  QueryKindDetail,
  QueryKindList,
} from "@/api/types";

// Re-exported so callers (e.g. the Save-as-kind dialog) can import the request
// and detail shapes alongside the client function from one module.
export type { CreateQueryKindRequest, QueryKindDetail };

// `GET /api/v1/query/kinds` — the catalogue of declarative query-kinds (WS-10)
// a panel can invoke by reverse-DNS name instead of pasting raw SQL. Read-only;
// the kind picker lists these and a kind-mode `QueryRequest` names one. The SQL
// stays server-side — only the descriptive surface (name, description,
// datasource shape, params schema) crosses the wire.
export function listQueryKinds(client: StarterClient): Promise<QueryKindList> {
  return fetchJson<QueryKindList>(client, `${client.apiPrefix}/query/kinds`, {
    method: "GET",
  });
}

// `POST /api/v1/query-kinds` (hyphen, NOT `/query/kinds`) — promote raw SQL the
// user just authored into a reusable, server-validated query-kind. The server
// LINTS the SQL on save: if `tables` is non-empty the SQL must filter by
// `$caller_tenant_id` (else 400), and any `$param` referenced must be declared
// in `params_schema` (else 400); a duplicate `name` returns 409. Those messages
// surface verbatim via the thrown `StarterError.message`.
export function createQueryKind(
  client: StarterClient,
  body: CreateQueryKindRequest,
): Promise<QueryKindDetail> {
  return fetchJson<QueryKindDetail>(client, `${client.apiPrefix}/query-kinds`, {
    method: "POST",
    headers: { "content-type": "application/json", ...readCsrfHeader() },
    body: JSON.stringify(body),
  });
}

File: nexus/backend/crates/nexus-spi/src/dto/query/kinds.rs
//! `GET /api/v1/query/kinds` — the catalogue of registered query-kinds (WS-10).
//!
//! The kind picker in the query editor reads this to list the declarative
//! queries a panel can invoke by name instead of pasting raw SQL. It exposes
//! only the descriptive surface (name, description, datasource shape, and the
//! params JSON Schema) — never the SQL, which stays server-side.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use utoipa::ToSchema;

/// The registered query-kinds, name-ordered, for the editor's kind picker.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct QueryKindList {
    /// Every registered kind, sorted by `name`.
    pub kinds: Vec<QueryKindSummary>,
}

/// One registered query-kind's descriptive surface. The `params_schema` is the
/// kind's JSON Schema document, which a schema-driven form renders to collect
/// the named params the kind binds.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct QueryKindSummary {
    /// Reverse-DNS id a `QueryRequest.kind` invokes (e.g. `nexus.core.meters_list`).
    pub name: String,
    /// Human description for the picker, if the manifest declared one.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// The datasource shape the kind targets (e.g. `postgres`).
    pub datasource_kind: String,
    /// The kind's params JSON Schema — the contract a kind-mode request's
    /// `params` validates against, and what a schema-driven form renders.
    pub params_schema: Value,
}
